windows/wrapper/impl_webrtc_AudioDeviceWasapi.h

```diff
@@ -0,0 +1,520 @@
+      void _TraceCOMError(HRESULT hr) const;
+      void _Get44kHzDrift();
+
+      int32_t _RefreshDeviceList(DeviceClass cls);
+      int16_t _DeviceListCount(DeviceClass cls);
+
+      winrt::hstring  _GetDefaultDeviceName(DeviceClass cls);
+      winrt::hstring  _GetListDeviceName(DeviceClass cls, int index);
+      winrt::hstring  _GetDeviceName(DeviceInformation const &device);
+
+      winrt::hstring  _GetListDeviceID(DeviceClass cls, int index);
+      winrt::hstring  _GetDefaultDeviceID(DeviceClass cls);
+      winrt::hstring  _GetDeviceID(DeviceInformation const &device);
+
+      DeviceInformation _GetDefaultDevice(DeviceClass cls, AudioDeviceRole role);
+      DeviceInformation _GetListDevice(DeviceClass cls, int index);
+      DeviceInformation _GetListDevice(DeviceClass cls, winrt::hstring const &deviceId);
+
+      void _InitializeAudioDeviceIn(winrt::hstring const &deviceId);
+      void _InitializeAudioDeviceOut(winrt::hstring const &deviceId);
+
+      // Surround system support
+      bool ShouldUpmix();
+      WAVEFORMATEX* GenerateMixFormatForMediaEngine(
+        WAVEFORMATEX* actualMixFormat);
+      WAVEFORMATPCMEX* GeneratePCMMixFormat(WAVEFORMATEX* actualMixFormat);
+      void Upmix(int16_t *inSamples, uint32_t numberOfFrames,
+        int16_t *outSamples, uint32_t inChannels, uint32_t outChannels);
+      void UpmixAndConvert(int16_t *inSamples, uint32_t numberOfFrames,
+        float *outSamples, uint32_t inChannels, uint32_t outChannels);
+
+   private:
+      void DefaultAudioCaptureDeviceChanged(
+        DefaultAudioCaptureDeviceChangedEventArgs const &args);
+      void DefaultAudioRenderDeviceChanged(
+        DefaultAudioRenderDeviceChangedEventArgs const &args);
+
+   private:
+     AudioDeviceWasapiWeakPtr thisWeak_;
+     mutable zsLib::RecursiveLock lock_;
+
+     IAudioDeviceWasapiDelegateSubscriptions subscriptions_;
+     IAudioDeviceWasapiSubscriptionPtr defaultSubscription_;
+
+     std::string id_;
+
+   private:
+      ScopedCOMInitializer                    _comInit;
+      AudioDeviceBuffer*                      _ptrAudioBuffer;
+      rtc::CriticalSection                    _critSect;
+      rtc::CriticalSection                    _volumeMutex;
+      rtc::CriticalSection                    _recordingControlMutex;
+      rtc::CriticalSection                    _playoutControlMutex;
+      int32_t                                 _id;
+
+   private:  // MMDevice
+      winrt::hstring      _deviceIdStringIn;
+      winrt::hstring      _deviceIdStringOut;
+      DeviceInformation   _captureDevice;
+      DeviceInformation   _renderDevice;
+
+      WAVEFORMATPCMEX*       _mixFormatSurroundOut;
+      bool                   _enableUpmix;
+
+      DeviceInformationCollection  _ptrCaptureCollection;
+      DeviceInformationCollection  _ptrRenderCollection;
+      DeviceInformationCollection  _ptrCollection;
+
+   private:  // WASAPI
+      IAudioClient*                           _ptrClientOut;
```

Comment:
Unless they are initialized in the constructor, these should be defaulted with {} to ensure they become NULL upon creation (true of all basic types). If they are only initialized in the constructor then best to move initialization to the header.

Reply:
All are initialized in constructor.